src/python/pants/engine/internals/scheduler.py

```diff
@@ -659,18 +659,15 @@ def register_task(rule: TaskRule) -> None:
         for selector in rule.input_selectors:
             native_engine.tasks_add_select(tasks, selector)
 
-        def add_get_edge(product: type, subject: type) -> None:
-            native_engine.tasks_add_get(tasks, product, subject)
-
         for the_get in rule.input_gets:
             if is_union(the_get.input_type):
-                # If the registered subject type is a union, add Get edges to all registered
-                # union members.
+                # Register a union. TODO: See #12934: this should involve an explicit interface
+                # soon, rather than one being implicitly created with only the provided Param.
                 for union_member in union_membership.get(the_get.input_type):
-                    add_get_edge(the_get.output_type, union_member)
+                    native_engine.tasks_add_union(tasks, the_get.output_type, (union_member,))
```

Comment:
It seems like it would now be more descriptive to just register the complete union:
```python
native_engine.tasks_add_union(tasks, the_get.output_type, tuple(union_membership.get(the_get.input_type)))
```
The TODO probably makes this not worth your while though.

Reply:
Good thought. And yea, can probably wait for #12934. Interested in your feedback on how we should specify plugin interfaces there!